src/digraph.rs

```diff
@@ -2265,6 +2274,159 @@ impl PyDiGraph {
+    /// :param object obj: The data/weight to associate with the new node.
+    /// :param bool check_cycle: If set to ``True``, validates
+    ///     that the substitution will not introduce cycles before
+    ///     modifying the graph. If set to ``False``, validation is
+    ///     skipped. If not provided, inherits the value
+    ///     of ``check_cycle`` from this instance of
+    ///     :class:`retworkx.PyDiGraph`.
+    /// :param weight_combo_fn: An optional python callable that, when
+    ///     specified, is used to merge parallel edges introduced by the
+    ///     substitution, which will occur when multiple nodes in
+    ///     ``to_replace`` have an incoming edge
+    ///     from the same source node or when multiple nodes in
+    ///     ``to_replace`` have an outgoing edge to the same target node.
+    ///     If this instance of :class:`retworkx.PyDiGraph` is a multigraph,
+    ///     leave this unspecified to preserve parallel edges. If unspecified
+    ///     when not a multigraph, parallel edges and their weights will be
+    ///     combined by choosing one of the edge's weights arbitrarily based
+    ///     on an internal iteration order, subject to change.
+    /// :returns: The index of the newly created node.
+    /// :raises DAGWouldCycle: The cycle check is enabled and the
+    ///     substitution would introduce cycle(s).
+    #[pyo3(
+        text_signature = "(self, to_replace, obj, /, check_cycle=None, weight_combo_fn=None)"
+    )]
+    fn substitute_nodes_with_node(
+        &mut self,
+        py: Python,
+        to_replace: Vec<usize>,
+        obj: PyObject,
+        check_cycle: Option<bool>,
+        weight_combo_fn: Option<PyObject>,
+    ) -> PyResult<usize> {
+        let can_contract = |nodes: &HashSet<NodeIndex>| {
+            // Start with successors of `nodes` that aren't in `nodes` itself.
+            let visit_next: Vec<NodeIndex> = nodes
+                .iter()
+                .flat_map(|n| self.graph.edges(*n))
+                .filter_map(|edge| {
+                    let target_node = edge.target();
+                    if !nodes.contains(&target_node) {
+                        Some(target_node)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            // Now, if we can reach any of `nodes`, there exists a path from `nodes`
+            // back to `nodes` of length > 1, meaning contraction is disallowed.
+            let mut dfs = Dfs::from_parts(visit_next, self.graph.visit_map());
+            while let Some(node) = dfs.next(&self.graph) {
+                if nodes.contains(&node) {
+                    // we found a path back to `nodes`
+                    return false;
+                }
+            }
+
+            true
+        };
+
+        let mut indices_to_remove: HashSet<NodeIndex> =
+            to_replace.iter().map(|&n| NodeIndex::new(n)).collect();
```

Comment:
```suggestion
        let mut indices_to_remove: indexmap::set::IndexSet<NodeIndex> =
            to_replace.iter().map(|&n| NodeIndex::new(n)).collect();
        indices_to_remove.sort()
```

Makes the code deterministic addressing https://github.com/Qiskit/retworkx/pull/487#pullrequestreview-811815333 without changing much of the code. `indexmap` is already a dependency we use extensively, might as well use it in this case we need determinism.

Reply:
That could work! Thoughts on using `collections::BTreeSet`? I've seen discussions of performance actually being better in some cases (better cache locality), and we can get an ordering without an explicit sort:

```rs
let mut indices_to_remove: BTreeSet<NodeIndex> =
  to_replace.into_iter().map(|n| NodeIndex::new(n)).collect();
```

I think we also need to fix `node_to_weight` as @georgios-ts mentioned as well to be deterministic. I am inclined to go for a `BTreeMap` in that case as well.

Reply:
Done in a45fd8b. I went with `indexmap`, but omitted the sort. We should still have determinism from the caller's perspective, but IMO there's not much reason to ensure that different orderings of `to_remove` yield the same result.